Cython/Compiler/Nodes.py

```diff
@@ -1821,6 +1821,10 @@ def generate_function_definitions(self, env, code):
 
         for entry in lenv.var_entries:
             if not (entry.in_closure or entry.is_arg):
+                if entry.type.is_memoryviewslice and not entry.init:
+                    # inferred memoryview slices don't get initialized anywhere otherwise
+                    # it's not clear where the best place to put this is
+                    entry.init = entry.type.default_value
```

Comment:
As you already guessed, this is not the best place to put this. ;)
`entry.init` is set in a couple of places, but usually at declaration time. Type inference runs after the `analyse_declarations()` phase, so that's too early for this case. Maybe `NameNode.analyse_expressions()` can do it? This is about variable entries, which should always be names.

Reply:
In `NameNode.analyse_declarations` it wasn't getting run there for the simple case `for xslice in x: pass` (`xslice` is apparently never analysed). I added it to type inference which I think is better.